syntax = "proto3";

package federated_learning;

service FederatedLearningService {
    // Client receives model parameters from Server and returns updated parameters
    rpc Fit (FitRequest) returns (FitResponse);
    
    // Optional: Evaluate the model on the client
    rpc Evaluate (EvaluateRequest) returns (EvaluateResponse);

    // New: Request a specific model architecture based on device capabilities
    rpc GetModelForDevice (GetModelRequest) returns (GetModelResponse);
}

message DeviceInfo {
    string device_type = 1;
    float memory_total_mb = 2;
    float memory_available_mb = 3;
    float compute_score = 4;
    int32 cpu_count = 5;
}

message FitRequest {
    // Model parameters serialized as bytes
    bytes parameters = 1; 
    // Configuration dictionary (serialized JSON or bytes) for hyperparameters
    map<string, string> config = 2;
    // Client device capabilities
    DeviceInfo device_info = 3;
}

message FitResponse {
    // Updated model parameters serialized as bytes
    bytes parameters = 1;
    // Number of examples used for training
    int32 num_examples = 2;
    // Metrics (e.g., accuracy, loss)
    map<string, float> metrics = 3;
}

message EvaluateRequest {
    bytes parameters = 1;
    map<string, string> config = 2;
}

message EvaluateResponse {
    float loss = 1;
    int32 num_examples = 2;
    map<string, float> metrics = 3;
}

message GetModelRequest {
    DeviceInfo device_info = 1;
}

message GetModelResponse {
    // Initial model parameters for the chosen complexity
    bytes parameters = 1;
    // Model configuration (e.g. architecture name)
    map<string, string> config = 2;
}
